### PR TITLE
Added support for guards when advancing workflow from a command

### PR DESCRIPTION
**Pull request by GDIBass**
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #23904
| License       | MIT
| Doc PR        | symfony/symfony-docs

Added support for advancing a workflow from the command line when the transition has guards.

### CODE REVIEW

File: src/Symfony/Component/Workflow/EventListener/GuardListener.php [hunk 1]
```diff
@@ -55,6 +56,11 @@ private function getVariables(GuardEvent $event)
     {
         $token = $this->tokenStorage->getToken();
 
+        if ($token == null) {
```
Comment (stloyd) [inline src/Symfony/Component/Workflow/EventListener/GuardListener.php]: `null === $token`

Comment (GDIBass) [inline src/Symfony/Component/Workflow/EventListener/GuardListener.php]: Updated.

**Comment by nicolas-grekas** [comment]
ping @Nyholm @lyrixx

**Comment by lyrixx** [comment]
Hello. 

@nicolas-grekas you can simply add the workflow label instead of pinging me ;) (Both are fine anyway)

@GDIBass:

I'm :-1: with the current implementation. 
It's the developer responsibility to set the right token when using the security from the CLI.

I understand your PR it's a convenient fallback but IMHO it will lead to many WTF for people that are not aware to see "feature"/fallaback.

May be we can instead throw an exception if the token is null. WDYT?

**Comment by GDIBass** [comment]
@lyrixx Hrmm, that makes sense.

In my own implementation, I'm only writing expressions that check the subject.   It would be nice if we could only throw an exception only if the expression required the user (like it used an is_granted check), but that would mean modifying the Expression language, right?

**Comment by lyrixx** [comment]
It is possible but it will need more work (I mean much more CPU work). Not sure it worth it.

**Comment by GDIBass** [comment]
How about something like this?

https://pastebin.com/34hgUi7j

**Comment by lyrixx** [comment]
it will lead to PHP warnings in the expression evaluation.

**Comment by GDIBass** [comment]
Ok, gotcha.  So I guess just something like this:

https://pastebin.com/cc9iazXs

**Comment by lyrixx** [comment]
Do you want to create a pull request ?

File: src/Symfony/Component/Workflow/EventListener/GuardListener.php [hunk 2]
```diff
@@ -55,6 +56,10 @@ private function getVariables(GuardEvent $event)
     {
         $token = $this->tokenStorage->getToken();
 
+        if (null === $token) {
+            throw new \Exception("No token is set");
```
Comment (stloyd) [inline src/Symfony/Component/Workflow/EventListener/GuardListener.php]: Please don't use top level exceptions.

File: src/Symfony/Component/Workflow/EventListener/GuardListener.php [hunk 3]
```diff
@@ -55,6 +55,10 @@ private function getVariables(GuardEvent $event)
     {
         $token = $this->tokenStorage->getToken();
 
+        if (null === $token) {
+            throw new \Exception('No token is set');
```
Comment (lyrixx) [inline src/Symfony/Component/Workflow/EventListener/GuardListener.php]: I don't know why @stloyd's comment is lost ; but he was right.
You should use an exception that lives in the Workflow component. If the existing exception does not match
this case, you will have to create a new one.

File: src/Symfony/Component/Workflow/Tests/EventListener/GuardListenerTest.php [hunk 4]
```diff
@@ -69,6 +69,16 @@ public function testWithSupportedEventAndAccept()
         $this->assertTrue($event->isBlocked());
     }
 
+    public function testWithNoTokenStorage()
+    {
+        $event = $this->createEvent();
+        $this->tokenStorage = null;
```
Comment (lyrixx) [inline src/Symfony/Component/Workflow/Tests/EventListener/GuardListenerTest.php]: this does not work.
You should configure the token storage to return `null`

Comment (GDIBass) [inline src/Symfony/Component/Workflow/Tests/EventListener/GuardListenerTest.php]: Cool, thanks for your patience here.  First time contributing!

I'll get this done right away

File: src/Symfony/Component/Workflow/EventListener/GuardListener.php [hunk 5]
```diff
@@ -55,6 +56,10 @@ private function getVariables(GuardEvent $event)
     {
         $token = $this->tokenStorage->getToken();
 
+        if (null === $token) {
+            throw new InvalidTokenConfigurationException('No token is set');
```
Comment (Simperfit) [inline src/Symfony/Component/Workflow/EventListener/GuardListener.php]: Could maybe be something like `sprintf('There are no token available for the transition event %s', $eventName)` ?

WDYT @lyrixx ?

**Comment by Simperfit** [comment]
Travis failure seems unrelated

File: src/Symfony/Component/Workflow/EventListener/GuardListener.php (hunk 5 above)
Comment (GDIBass) [reply src/Symfony/Component/Workflow/EventListener/GuardListener.php]: So: 

`throw new InvalidTokenConfigurationException(sprintf('There are no token available for the transition event %s', $event->getWorkflowName()));`

**Review by Simperfit — APPROVED** [review]

File: src/Symfony/Component/Workflow/EventListener/GuardListener.php [hunk 6]
```diff
@@ -55,6 +56,10 @@ private function getVariables(GuardEvent $event)
     {
         $token = $this->tokenStorage->getToken();
 
+        if (null === $token) {
+            throw new InvalidTokenConfigurationException(sprintf('There are no token available for workflow %s', $event->getWorkflowName()));
```
Comment (xabbuh) [inline src/Symfony/Component/Workflow/EventListener/GuardListener.php]: [...] no tokens [...]

Comment (xabbuh) [reply src/Symfony/Component/Workflow/EventListener/GuardListener.php]: and the exception message should end with a dot

File: src/Symfony/Component/Workflow/Tests/EventListener/GuardListenerTest.php [hunk 7]
```diff
@@ -69,6 +69,18 @@ public function testWithSupportedEventAndAccept()
         $this->assertTrue($event->isBlocked());
     }
 
+    /**
+     * @expectedException \Symfony\Component\Workflow\Exception\InvalidTokenConfigurationException
+     * @expectedExceptionMessage There are no token available for workflow unnamed
+     */
+    public function testWithNoTokenStorage()
```
Comment (xabbuh) [inline src/Symfony/Component/Workflow/Tests/EventListener/GuardListenerTest.php]: actually, there is a token storage but no token

File: src/Symfony/Component/Workflow/EventListener/GuardListener.php (hunk 6 above)
Comment (GDIBass) [reply src/Symfony/Component/Workflow/EventListener/GuardListener.php]: Updated.

File: src/Symfony/Component/Workflow/Tests/EventListener/GuardListenerTest.php (hunk 7 above)
Comment (GDIBass) [reply src/Symfony/Component/Workflow/Tests/EventListener/GuardListenerTest.php]: Updated!

**Comment by lyrixx** [comment]
:+1:

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @GDIBass.

**Pull request merged**